### ISSUE TITLE
Add ExclusiveResource programmatically

### ISSUE THREAD

**Issue by U1F984** [junit-team/junit5#2677]
I have several test classes, each with multiple test methods inside. Some of those classes are injected with a parameter, and when running classes in parallel those classes should not run parallel. If I understand correctly, the `@ResourceLock` annotation can help here, but it would be very inconvenient to manually add that annotation to hundereds of tests. Ideally a extension could take a look at the class, see that it has this parameter and somehow add the ExclusiveResource to it.

My JUnit execution parameters:

```
-Djunit.jupiter.execution.parallel.enabled=true
-Djunit.jupiter.execution.parallel.mode.default=same_thread
-Djunit.jupiter.execution.parallel.mode.classes.default=concurrent
-Djunit.jupiter.execution.parallel.config.strategy=fixed
-Djunit.jupiter.execution.parallel.config.fixed.parallelism=8
```

**Comment by marcphilipp** [comment]
We could add a new extension point for this:

```java
interface ResourceLockProvider extends Extension {
	default Set<ExclusiveResource> provideResourcesForClass(Class<?> testClass, Set<ExclusiveResource> declaredResources) {
		return declaredResources;
	};
	Set<ExclusiveResource> provideResourcesForNestedClass(Class<?> nestedClass, Set<ExclusiveResource> declaredResources) {
		return declaredResources;
	};
	Set<ExclusiveResource> provideResourcesForMethod(Class<?> testClass, Method testMethod, Set<ExclusiveResource> declaredResources) {
		return declaredResources;
	};
}
interface ExclusiveResource {
	String getKey();
	ResourceAccessMode getAccessMode();
	static ExclusiveResource of(String key);
	static ExclusiveResource of(String key, ResourceAccessMode accessMode);
}

```

**Comment by jbduncan** [comment]
This feature might be quite useful for my [PR](https://github.com/junit-pioneer/junit-pioneer/pull/491) on junit-pioneer to allow "resources" - arbitrary objects - to be instantiated and shared amongst tests declaratively. 

Specifically, if each "resource" were to have its own `ResourceLock` that could be created programmatically, then it would simplify making my extension thread-safe.

The proposed API leaves me with one question, though: if my extension were to implement `ResourceLockProvider`, when would the `provideX` methods be called?

**Comment by marcphilipp** [comment]
It would have to be called immediately before starting execution on the engine level.

**Comment by jbduncan** [comment]
Okay, thank you!

I wasn't sure at first if this would work for me, since my extension's resources are created at test runtime. But I think it could now, as all shared resources in my extension are declared with `@Shared` annotations, which I think my potential `ResourceLockProvider` could search for reflectively.

So I'm happy with the proposed API. 👍

**Comment by marcphilipp** [comment]
**Team decision:** Let's go with the above proposal.

@U1F984 @jbduncan Is one of you interested in working on this?

**Comment by jbduncan** [comment]
@marcphilipp I'd love to! However I still have https://github.com/junit-pioneer/junit-pioneer/issues/348 to finish, so I'd be happy for anyone else to pick this up before me.

**Comment by jbduncan** [comment]
For clarity, my JUnit Pioneer work is not actually blocked on this feature, as I've found an alternative solution to programmatic `ResourceLock`s for my needs.

**Comment by jbduncan** [comment]
@marcphilipp A follow-up query from me: do you know if JUnit Jupiter already ensures that `ResourceLock`s are obtained in such an order as to prevent the [dining philosophers problem](https://en.wikipedia.org/wiki/Dining_philosophers_problem), by any chance?

If not, then I, or whoever picks this up, will need to ensure the locks are sorted when being moved from `ResourceLockProvider`s to the rest of the test framework, by some definition of "sorted".

**Comment by marcphilipp** [comment]
They are sorted here: https://github.com/junit-team/junit5/blob/6a47ae152003a6ac3e77f2e604159b71e4f82953/junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/LockManager.java#L61

**Comment by jbduncan** [comment]
@marcphilipp Great, thank you for answering so quickly. 👍

**Comment by jbduncan** [comment]
@marcphilipp I'd love it if you assigned this to me, as I've managed to park my junit-pioneer work and find time to start looking at this.

**Comment by marcphilipp** [comment]
@jbduncan Sorry it took 6 days, but you're now assigned.

**Comment by jbduncan** [comment]
Sorry for the silence I've this! I was a bit lost looking through the JUnit 5 codebase when I looked at this issue last year. But now that my [JUnit Pioneer PR](https://github.com/junit-pioneer/junit-pioneer/pull/491) that would have used this feature is finally done, I'll see if I can revisit this issue later this week, and write down anything I need help understanding. :)

**Comment by jbduncan** [comment]
@marcphilipp I have a few follow-up questions which I was hoping you could help me with.

For the proposed `ExclusiveResourceProvider`,

1. Does a `provide*` method's `declaredResources` contain just the resources found so far for the given class/method/nested class, or does it contain _all_ resources found so far by JUnit?
2. Do the `declaredResources` contain resources found through `@ResourceLock`s?
    ```java
    @Test
    @ResourceLock(value = "a")
    @ExtendWith(MyExclusiveResourceProvider.class)
    void test() {
    }
    ```

3.  If so, does the ordering matter? As in, does `MyExclusiveResourceProvider`'s declared resources contain resource "a" when running with `test2`, but not with `test1`?
    ```java
    @Test
    @ExtendWith(MyExclusiveResourceProvider.class)
    @ResourceLock(value = "a")
    void test1() {
    }

    @Test
    @ResourceLock(value = "a")
    @ExtendWith(MyExclusiveResourceProvider.class)
    void test2() {
    }

    class MyExclusiveResourceProvider implements ExclusiveResourceProvider {
        @Override public Set<ExclusiveResource> provideExclusiveResourcesForMethod(Class<?> testClass, Method testMethod, Set<ExclusiveResource> declaredResources) {
            return declaredResources;
        }
    }
    ```

**Comment by jbduncan** [comment]
I've pushed my ongoing spike to https://github.com/junit-team/junit5/pull/3096 for the sake of transparency. :)

**Comment by jbduncan** [comment]
@marcphilipp I'm feeling a bit stuck now, and I was wondering if you could help me.

It looks like `NodeTreeWalker` is responsible for discovering tests and, most importantly, their `@ResourceLock`s. However, to get resource locks from `ExclusiveResourceProvider`s, JUnit would need to discover them as extensions, which are currently discovered by `NodeTestTask`s, which in turn depend on the `NodeTreeWalker`! So there's a cycle of dependencies (maybe even a web), and I can't see how to break it, so I was was wondering if you had any thoughts?

Alternatively, would you be up to discussing this remotely some time? If so, feel free to contact me on my GitHub email address.

**Comment by jbduncan** [comment]
@marcphilipp I don't suppose you've had the time to look at my last comment recently, have you? 🙂

**Comment by marcphilipp** [comment]
@jbduncan Sorry, I had fallen quite behind in reading my GitHub notifications. 😬 

> It looks like `NodeTreeWalker` is responsible for discovering tests and, most importantly, their `@ResourceLock`s. However, to get resource locks from `ExclusiveResourceProvider`s, JUnit would need to discover them as extensions, which are currently discovered by `NodeTestTask`s, which in turn depend on the `NodeTreeWalker`!

I'm not sure I fully understand the question. `ExclusiveResourceProvider` can't be a regular extension since it needs to be applied at discovery time, i.e. it should not extend `Extension`. We already have a few of this kind of extension, e.g. `DisplayNameGenerator` and `MethodOrderer`. So I think we'll need a new annotation that goes along with `ExclusiveResourceProvider`, maybe `@ExclusiveResources(MyExclusiveResourceProvider.class)`?

**Comment by jbduncan** [comment]
> @jbduncan Sorry, I had fallen quite behind in reading my GitHub notifications. 😬

No worries! Likewise, sorry for not responding immediately. :)

> I'm not sure I fully understand the question. `ExclusiveResourceProvider` can't be a regular extension since it needs to be applied at discovery time, i.e. it should not extend `Extension`. We already have a few of this kind of extension, e.g. `DisplayNameGenerator` and `MethodOrderer`. So I think we'll need a new annotation that goes along with `ExclusiveResourceProvider`, maybe `@ExclusiveResources(MyExclusiveResourceProvider.class)`?

@marcphilipp Oh, I hadn't considered that. That sounds like it could break the dependency cycle.

But admittedly I've lost interest in making programmatic exclusive resources a reality, so I'll leave this issue and my [draft PR](https://github.com/junit-team/junit5/pull/3096) for someone else who in interested in picking things up.

**Comment by marcphilipp** [comment]
No worries. Thanks for letting us know!

**Comment by VladimirDmitrienko** [comment]
Hi @marcphilipp 

Me and my team interested in this feature. 
We have a lot of tests and would like to configure locking strategy without adding / removing `@ResourceLock ` annotation.

I've prepared a draft implementation of this functionality and would be very grateful if you take a look at it:
https://github.com/junit-team/junit5/pull/3889

Regards,
Vladimir

**Comment by marcphilipp** [comment]
@VladimirDmitrienko Thanks for the draft PR! Could you please describe your use case in more detail? Is a class-level annotation sufficient or would you rather register such a provider "globally" for all tests?

**Comment by VladimirDmitrienko** [comment]
@marcphilipp Thanks for a quick response!

We have lots of tests which interact with our application through different levels: UI, API, DB.
When we run our tests in parallel, sometimes they may "intersect" because a test or a group of tests may modify a common state used by other tests (a typical problem actually).

My initial thought was to use `@ExclusiveResources(MyExclusiveResourceProvider.class)` on the base test class and assign different locks right from `MyExclusiveResourceProvider`. 
In such case we would be able to add a necessary level of isolation between tests based on class / package names patterns. 
Making sure that **particular kinds / groups of tests are never executed at the same time but still executed in a "concurrent" manner.**

---

So, responding to your question, the current approach would be sufficient for our case, but I guess that a "global" `ExclusiveResourceProvider` may be an even better solution since it will not require users to have a "base" class for all the tests.
For such purpose we could introduce a new property, maybe something like:
```properties
junit.jupiter.execution.parallel.locks.provider = org.com.example.MyExclusiveResourceProvider
```

---

To summarize, currently I see three options to set `ExclusiveResourceProvider`:

**1.** `@ExclusiveResources` annotation.
In such case I suppose it's better to limit it's target to classes:
```java
@Target({ ElementType.TYPE}`)
```

**2.** `junit.jupiter.execution.parallel.locks.provider` property.

**3.** Combine both approaches to be able to override the property with an annotation.
Looks like the most flexible since fits both: 
small-medium size projects which can benefit from having all the logic "in one place";
big projects with huge test classes hierarchy: each `Base...Tests` class can have it's own `ExclusiveResourceProvider`.

So, I'm looking forward to your feedback to continue to work on whichever approach we choose.

**Comment by VladimirDmitrienko** [comment]
@marcphilipp maybe you had a chance to take a look at my last comment?

Also, I'd appreciate if you assign this issue on me 🙂

**Comment by marcphilipp** [comment]
I think we should go with a class-level annotation for now. We can always add a config parameter or `ServiceLoader`-based registration later. Let's call it `@ResourceLocks` and `ResourceLockProvider`, though, to stick to the terminology used in the Jupiter API.

**Comment by VladimirDmitrienko** [comment]
@marcphilipp got your point.
I'm afraid `@ResourceLocks` could be a bit confusing since we already have such annotation. Don't you think?

**Comment by marcphilipp** [comment]
You're right, I forgot about the container annotation. I think `@ResourceLockProvider` would be good but then I don't know what we'd call the interface. Any ideas?

**Comment by VladimirDmitrienko** [comment]
What do you think about:
```java
@ResourceLocksSource(MyResourceLockProvider.class)
```
The same approach is used in `@ArgumentsSource`.

**Comment by marcphilipp** [comment]
I thought about that as well but concluded that it might be confusing since it's unrelated to the other source annotations.

**Comment by VladimirDmitrienko** [comment]
Fair enough.

There are two options coming in my mind.
1. What if we reuse existing `@ResourceLocks` annotation:
```java
public @interface ResourceLocks {

	ResourceLock[] value();

	Class<? extends ResourceLocksProvider> provider() default DefaultResourceLocksProvider.class;
}
```
2. `@ResourceLocksFrom(MyResourceLocksProvider.class)`

**Comment by VladimirDmitrienko** [comment]
Agreed 👍 
As always, naming is hard 😅

Just curious, what kind of compatibility will it break?
I supposed adding a new method will be both a source and binary compatible change.

**Comment by marcphilipp** [comment]
Did you mean adding another attribute to the container annotation? If so, you may be right but I still think it would be confusing to appropriate it in this way.

**Comment by VladimirDmitrienko** [comment]
Got it, thank you 🙂 
Will keep you updated on my work.

**Comment by VladimirDmitrienko** [comment]
Hi @marcphilipp 

I've updated the [PR](https://github.com/junit-team/junit5/pull/3889) and now it's ready for review.
Looking forward to your feedback.

Btw, I've set `API since` to `5.12` but actually I'm not really sure in which release it will go, please, let me know and I'll update `since` values and add release notes.

**Comment by VladimirDmitrienko** [comment]
Hi @marcphilipp 
I would be happy to improve my [PR](https://github.com/junit-team/junit5/pull/3889) and waiting for a feedback from your side.
Maybe you have any at least approximate estimates of when could I expect that?
Would be glad to receive any news on this one.

**Comment by marcphilipp** [comment]
Sorry for the delay! I was on vacation. I should be able to get to it this week.

**Comment by VladimirDmitrienko** [comment]
@marcphilipp Thanks for the update 👍

**Comment by marcphilipp** [comment]
Reopening to track polishing work

**Comment by nickh-stripe** [comment]
little late to the party here, looking at the new experimental api for this: https://junit.org/junit5/docs/snapshot/api/org.junit.jupiter.api/org/junit/jupiter/api/parallel/ResourceLocksProvider.html

was it purposeful to exclude the test template / parameterised tests? seems you can only provide dynamically for class/method and nested class and not a templated or parameterised test which is what I was hoping for for my usecase..

**Comment by VladimirDmitrienko** [comment]
Hi @nickh-stripe 

Actually `ResourceLocksProvider` can be used to define locks for parameterized tests.
However, the same locks will be applied to each invocation of the test, _regardless of the test input_.

Let's consider such an example:
```java
	@ParameterizedTest
	@ValueSource(strings = {"banana", "apple", "orange"})
	void fruitTest(String value) {
		// ...
	}

	static class Provider implements ResourceLocksProvider {

		@Override
		public Set<Lock> provideForMethod(Class<?> testClass, Method testMethod) {
			return Set.of(new Lock("fruitKey"));
		}
	}
```
The same lock applies to each invocation: `fruitTest("banana")`, `fruitTest("apple")`, `fruitTest("orange")`.

It's impossible to define different locks based on the arguments, for example, setting the same lock for "banana" and "apple" but a different one for "orange". The reason is that _locks are collected before test execution started_: we first collect the locks and only then pass them to engine that executes tests.

----------------

Assigning locks depending on arguments could be achieved by splitting the test:  
```java
	@ParameterizedTest
	@ValueSource(strings = {"banana", "apple"})
	void bananaAndAppleTest(String value) {
		// ...
	}

	@ParameterizedTest
	@ValueSource(strings = {"orange"})
	void orangeTest(String value) {
		// ...
	}

	static class Provider implements ResourceLocksProvider {

		@Override
		public Set<Lock> provideForMethod(Class<?> testClass, Method testMethod) {
			if (testMethod.getName().equals("bananaAndAppleTest")) {
				return Set.of(new Lock("bananaAndAppleKey"));
			}
			else if (testMethod.getName().equals("orangeTest")) {
				return Set.of(new Lock("orangeKey"));
			}
			return Set.of();
		}
	}
```

**Comment by nickh-stripe** [comment]
thanks for the detailed explanation! the approach makes sense but is not possible for the shape of our problem unfortunately, splitting the tests based on their locks isn't really feasible unfortunately so i'll have to think about a different way to achieve it. 👍

**Comment by nickh-stripe** [comment]
i think this per method approach could work for me in the end, i just realised this is only in 5.12 tho, not in a 5.11.x release, any ideas on when 5.12 out? or this feature available in 5.11? 👍

**Comment by sbrannen** [comment]
Hi @nickh-stripe,

> i just realised this is only in 5.12 tho, not in a 5.11.x release, any ideas on when 5.12 out? or this feature available in 5.11? 👍

This feature is coming in JUnit Jupiter 5.12.

The milestone had unfortunately not be assigned for this issue, but I've just corrected that.

As you can see on the right as well as on the milestone page, 5.12 M1 is scheduled to be released by the end of January 2025.

An M1 is typically followed by zero or more additional milestones and at least one RC (release candidate), and the 5.12 GA release will follow that. So, if I had to guess, I'd say we will likely release 5.12 in the next 2-4 months.

**Comment by nickh-stripe** [comment]
thanks for the info, M1 by end of Jan works pretty well for my timelines i think! As long as its own maven central it should be easy to try out, i see from previous Mx versions that they were published to central (for people to test out i suppose!) so should be gtg. 👍
